packages/mobx-state-tree/src/types/utility-types/snapshotProcessor.ts

```diff
@@ -0,0 +1,210 @@
+    BaseType,
+    isType,
+    isStateTreeNode,
+    IValidationContext,
+    IValidationResult,
+    AnyObjectNode,
+    RedefineIStateTreeNode,
+    IStateTreeNode,
+    TypeFlags,
+    ExtractNodeType
+} from "../../internal"
+
+/** @hidden */
+declare const $mstNotCustomized: unique symbol
+
+/** @hidden */
+// tslint:disable-next-line:class-name
+export interface _NotCustomized {
+    // only for typings
+    readonly [$mstNotCustomized]: undefined
+}
+/** @hidden */
+export type _CustomOrOther<Custom, Other> = Custom extends _NotCustomized ? Other : Custom
+
+/** @hidden */
+export type SnapshotProcessorT<IT extends IAnyType, CustomC, CustomS> = RedefineIStateTreeNode<
+    ExtractT<IT>,
+    IStateTreeNode<_CustomOrOther<CustomC, ExtractC<IT>>, _CustomOrOther<CustomS, ExtractS<IT>>>
+>
+
+class SnapshotProcessor<IT extends IAnyType, CustomC, CustomS> extends BaseType<
+    _CustomOrOther<CustomC, ExtractC<IT>>,
+    _CustomOrOther<CustomS, ExtractS<IT>>,
+    SnapshotProcessorT<IT, CustomC, CustomS>,
+    ExtractNodeType<IT>
+> {
+    get flags() {
+        return this._subtype.flags | TypeFlags.SnapshotProcessor
+    }
+
+    constructor(
+        private readonly _subtype: IT,
+        private readonly _processors: ISnapshotProcessors<
+            ExtractC<IT>,
+            CustomC,
+            ExtractS<IT>,
+            CustomS
+        >,
+        name?: string
+    ) {
+        super(name || _subtype.name)
+    }
+
+    describe() {
+        return `snapshotProcessor(${this._subtype.describe()})`
+    }
+
+    private preProcessSnapshot(sn: this["C"]): ExtractC<IT> {
+        if (this._processors.preProcessor) {
+            return this._processors.preProcessor.call(null, sn)
+        }
+        return sn as any
+    }
+
+    private postProcessSnapshot(sn: ExtractS<IT>): this["S"] {
+        if (this._processors.postProcessor) {
+            return this._processors.postProcessor.call(null, sn) as any
+        }
+        return sn
+    }
+
+    instantiate(
+        parent: AnyObjectNode | null,
+        subpath: string,
+        environment: any,
+        initialValue: this["C"] | this["T"]
+    ): this["N"] {
+        const processedInitialValue = isStateTreeNode(initialValue)
+            ? initialValue
+            : this.preProcessSnapshot(initialValue)
+        const node = this._subtype.instantiate(parent, subpath, environment, processedInitialValue)
+        const oldGetSnapshot = node.getSnapshot
+        node.getSnapshot = () => {
+            return this.postProcessSnapshot(oldGetSnapshot.call(node)) as any
+        }
+        return node
+    }
+
+    reconcile(current: this["N"], newValue: this["C"] | this["T"]): this["N"] {
+        const node = this._subtype.reconcile(
+            current,
+            isStateTreeNode(newValue) ? newValue : this.preProcessSnapshot(newValue)
+        )
+        if (node !== current) {
+            const oldGetSnapshot = node.getSnapshot
+            node.getSnapshot = () => {
```

Comment:
I can imagine there is a quite unlikely edge case to run into, suppose you have

```ts
const Store = types.model({
  a: types.array(types.snapshotPreprocessor(Task)),
  b: types.array(Task)
})
```

What would happen if a Task is moved from `a` to `b` and vice versa? Since the nodes stay alive, the `getSnapshot` override is not (un)done, correct? Or is it forbidden anyway because `Task` is not assignable to `snapshotProcessor(Task)`?

Reply:
you mean with an actual processor that does something or with no processors at all? I can create a unit test to see what would happen.
What is the wanted behavior when a) the snapshot processor does some change and b) when it doesn't?

Reply:
@mweststrate I see now what you mean. Added unit tests and disallowed assignations unless they are of the exact same "snapshot processor" type.

Had to do it in a kind of "hackish" way, but it works.

That method could also be applied to optionals in case optionals should not be assignable to non-optionals (or vice versa)